### ISSUE TITLE
How to pass a custom mask to SAM for more precise results?

### ISSUE THREAD

**Issue by cip8** [facebookresearch/segment-anything#360]
Hi all 👋

### - custom SAM mask
I'm interested in passing a custom mask to `mask_input` - so not the logits generated previously by another SAM run, but a custom grayscale mask that I have.

Anyone knows how to accurately convert a custom grayscale or binary mask to SAM?

I'm attempting a conversion with the code below, but the results seem to be unchanged:

```
def reference_to_sam_mask(self, ref_mask: np.ndarray) -> np.ndarray:
    ref_mask = (ref_mask > 0).astype(np.uint8)

    # Resize to have the longest side 256
    height, width = ref_mask.shape[:2]
    new_height, new_width = (
        height * 256 // max(height, width),
        width * 256 // max(height, width),
    )

    ref_mask = cv2.resize(
        ref_mask, (new_width, new_height), interpolation=cv2.INTER_AREA
    )

    # Calculate padding size & pad mask evenly on all sides.
    pad_height = max(0, 256 - ref_mask.shape[0])
    pad_width = max(0, 256 - ref_mask.shape[1])
    pad_top = pad_height // 2
    pad_bottom = pad_height - pad_top
    pad_left = pad_width // 2
    pad_right = pad_width - pad_left

    ref_mask = np.pad(
        ref_mask, ((pad_top, pad_bottom), (pad_left, pad_right)), mode="constant"
    )

    return np.expand_dims(ref_mask, axis=0)  # expand to 1xHxW
```

### - mask_input's role
Also, do I understand the role of `**mask_input**` correctly? 
I assumed it can be used as a selector for SAM in the same as a **bounding box** or **input point**, but maybe I'm wrong - can someone with more insight bring some light into this please?

Thanks a ton! 💙🖖

**Comment by markushunter** [comment]
👍

**Comment by markushunter** [comment]
Similar thread: https://github.com/facebookresearch/segment-anything/issues/169

**Comment by markushunter** [comment]
I think you need to pad on the lower right corner and change the mask to different values: https://github.com/facebookresearch/segment-anything/issues/169#issuecomment-1550222757

**Comment by cip8** [comment]
Closing, as the issue is similar to #169 and #242 - following the discussions in those threads instead.

**Comment by wm-Githuber** [comment]
Hi, I encounter the same issue. First, I inputed the general mask into prompt_encoder, and scale the resuting mask_embedding by a factor of four. Then, the mask_embedding and Image_embedding are inputted into mask_decoder together. however, the Prediction is all black (no sense). Could you guide me?
